examples/vision/siamese.py

```diff
@@ -0,0 +1,290 @@
+
+"""
+# Siamese network with a contrastive loss
+
+Author: [Mehdi](https://github.com/s-mrb)<br>
+Date created: 2021/05/06<br>
+Last modified: 2020/05/06<br>
+Description: Similarity learning using siamese network with contrastive loss
+"""
+
+"""
+## Introduction
+
+[Siamese Network](https://en.wikipedia.org/wiki/Siamese_neural_network)
+is any Neural Network which share weights between two or more sister networks,
+each producing embedding vector of its respective input and these embeddings
+are then passed through some 
+[distance heuristic](https://developers.google.com/machine-learning/clustering/similarity/measuring-similarity)
+to find the distance between them. This distance is later used to increase the
+contrast between embeddings of inputs of different classes and decrease it with
+that of similar class by employing some loss function, with the main objective
+of contrasting [vector spaces](https://en.wikipedia.org/wiki/Vector_space)
+from which these sample inputs were taken.
+
+"""
+
+"""
+## Setup
+"""
+
+import random
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers.experimental.preprocessing import Rescaling
+from tensorflow.keras.layers import Flatten, Dense, Concatenate, Lambda, Input
+from tensorflow.keras.layers import Conv2D, Activation,AveragePooling2D
+from tensorflow.keras.datasets import mnist
+from tensorflow.keras import utils
+import matplotlib.pyplot as plt
+from keras import backend as K
+
+"""
+## Load the MNIST dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+
+# Change the data type to a floating point format
+x_train = x_train.astype('float32')
+x_test = x_test.astype('float32')
+
+"""
+## Create pairs of images
+
+We will train the model to differentiate each digit from one another. For
+example, digit `0` needs to be differentiated from the rest of the
+digits (`1` through `9`), digit `1` - from `0` and `2` through `9`, and so on.
+To carry this out, we will select N random images from class A (for example, for
+digit `0`) and pair it with N random images from another class B (for example,
+for digit `1`). Then, we can repeat this process for all classes of digits
+(until digit `9`). Once we have paired digit `0` with other digits, we can
+repeat this process for the remaining classes for the rest of the digits (from
+`1` until `9`).
+
+"""
+
+def make_pairs(x, y):
+    """
+    Parameters
+    ----------
+    x : list/array
+        List containing images, each index in this list corresponds to one image
+
+    y : list
+        List containing labels, each label with datatype of `int`
+
+    Returns
+    -------
+    List of the shape (number_of_pairs, shape_of_one_pair)
```

Comment:
You are returning tuple, not List. I'd suggest the following:
Returns: tuple containing two numpy arrays as (pair_of_samples, labels), where pair of samples' shape is (2*len(x), 2,n_features_dims) and labels are a binary array of shape (2*len(x))